miner/remote_proposals.go

```diff
@@ -0,0 +1,219 @@
+
+	clock     layerClock
+	publisher pubsub.Publisher
+	nodeSvc   nodeService
+	signers   struct {
+		mu      sync.Mutex
+		signers map[types.NodeID]*signerSession
+	}
+	epochEligibilities map[types.EpochID]uint32
+}
+
+// New creates a struct of block builder type.
+func NewRemoteBuilder(
+	clock layerClock,
+	publisher pubsub.Publisher,
+	svc nodeService,
+	layerSize uint32,
+	layersPerEpoch uint32,
+	logger *zap.Logger,
+) *RemoteProposalBuilder {
+	pb := &RemoteProposalBuilder{
+		cfg: config{
+			workersLimit:   runtime.NumCPU(),
+			activeSet:      DefaultActiveSetPreparation(),
+			layerSize:      layerSize,
+			layersPerEpoch: layersPerEpoch,
+		},
+		logger:             logger,
+		clock:              clock,
+		publisher:          publisher,
+		nodeSvc:            svc,
+		epochEligibilities: make(map[types.EpochID]uint32),
+		signers: struct {
+			mu      sync.Mutex
+			signers map[types.NodeID]*signerSession
+		}{
+			signers: map[types.NodeID]*signerSession{},
+		},
+	}
+	if logger == nil {
+		pb.logger = zap.NewNop()
+	}
+	return pb
+}
+
+func (pb *RemoteProposalBuilder) Register(sig *signing.EdSigner) {
+	pb.signers.mu.Lock()
+	defer pb.signers.mu.Unlock()
+	_, exist := pb.signers.signers[sig.NodeID()]
+	if !exist {
+		pb.logger.Info("registered signing key", log.ZShortStringer("id", sig.NodeID()))
+		pb.signers.signers[sig.NodeID()] = &signerSession{
+			signer: sig,
+			log:    pb.logger.With(zap.String("signer", sig.NodeID().ShortString())),
+		}
+	}
+}
+
+// Start the loop that listens to layers and build proposals.
+func (pb *RemoteProposalBuilder) Run(ctx context.Context) error {
+	current := pb.clock.CurrentLayer()
+	next := current + 1
+	pb.logger.Info("started", zap.Inline(&pb.cfg), zap.Uint32("next", next.Uint32()))
+	var eg errgroup.Group
+	prepareDisabled := pb.cfg.activeSet.Tries == 0 || pb.cfg.activeSet.RetryInterval == 0
+	if prepareDisabled {
+		pb.logger.Warn("activeset will not be prepared in advance")
+	}
+	for {
+		select {
+		case <-ctx.Done():
+			eg.Wait()
+			return nil
+		case <-pb.clock.AwaitLayer(next):
+			current := pb.clock.CurrentLayer()
+			if current.Before(next) {
+				pb.logger.Info("time sync detected, realigning ProposalBuilder",
+					zap.Uint32("current", current.Uint32()),
+					zap.Uint32("next", next.Uint32()),
+				)
+				continue
+			}
+			next = current.Add(1)
+			ctx := log.WithNewSessionID(ctx)
+
+			if current <= types.GetEffectiveGenesis() {
+				continue
+			}
+			if err := pb.build(ctx, current); err != nil {
+				pb.logger.Warn("failed to build proposal",
+					log.ZContext(ctx),
+					zap.Uint32("lid", current.Uint32()),
+					zap.Error(err),
+				)
+			}
+
+			pb.clean(current)
+		}
+	}
+}
+
+func (pb *RemoteProposalBuilder) clean(layer types.LayerID) {
+	var vals []types.EpochID
+	lim := layer.GetEpoch() - 2
+	for k := range pb.epochEligibilities {
+		if k <= lim {
+			vals = append(vals, k)
+		}
+	}
+	for _, v := range vals {
+		delete(pb.epochEligibilities, v)
+	}
```

Comment:
Why keep the last 2 epochs? I _think_ `pb.epochEligibilies` doesn't even need to be a field of `RemoteProposalBuilder` - it could be a local variable in the scope of the `RemoteProposalBuilder::Run` `for` loop, cleared on epoch boundary. The `clean()` method wouldn't be even needed and the scope of `epochEligibilies` and its usage would be much clearer to understand. Wdyt?

Reply:
technically you're right. and yes it is probably useless to keep the last 2 epochs.
the problem with calculating the eligibilities is, that right now we rely on the node to tell us how many times we're eligible in an epoch and if that's bigger than one we calculate the proofs. we also need the nonce (which i earlier said i delegated to the node service because of unclear persistence guarantees in the db).
if we sort out those dependencies to be somehow local, we could fetch the beacon once at the beginning of the epoch and then do the calculation once.

Reply:
I was thinking about:
1. define local `var eligibilities map[types.NodeID]map[types.LayerID][]types.VotingEligibility` in `RemoteProposalBuilder::Run`
2. on a new layer initialize eligibilities map for this epoch `if current.FirstInEpoch()`
3. move the `for _, signer := range signers` loop to the `Run()` method and call `build(ctx, layer, signer, proofs)` for each signer only if it is eligible in the `current` layer
4. remove `clean()`

There is less shared state (in the `RemoteProposalBuilder` struct) and the behaviour is more localized this way, which makes the code easier to understand and follow.

Wdyt?

Reply:
hmmm ok. but this is not just a simple reshuffle of logic, i need to add more logic and more api calls that move the whole logic of calculating the epoch eligibilities to the smesher service. right now this isn't the case, and as i mentioned, we're still relying on the remote node to tell us how many slots we're eligible for in this epoch. i agree that what you're suggesting is the right way to do it. if the scope of change sounds right to you, please confirm, and i'll do it this way.

Reply:
The suggested change is entirely enclosed in the miner/remote_proposals.go - it's just moving the code around. You don't need to add any API, etc.

Reply:
have a look at the inputs to the function that calculates the eligibility proofs:
```go
		proofs = calcEligibilityProofs(
					signer.signer.VRFSigner(),
					epoch,
					bcn,
					types.VRFPostIndex(nonce),
					proposal.Ballot.EpochData.EligibilityCount,
					pb.cfg.layersPerEpoch,
				)
```
it needs: the beacon value, the atx nonce, _the number of slots we're eligible for_. all of those are values that come from api calls. and you _must know_ how many times you're eligible for in this epoch. this comes from the partially built proposal that the node service returns. if we want to calculate the number of eligibilities on the smesher side too (fine, that's the thing we probably wanna do anyway), we must move the logic which is currently in `initSignerDataFor`(which is currently on the node-service side), which gets all the relevant data needed for the eligibility count calculation and does the calculation of number of slots.

> call `build(ctx, layer, signer, proofs)` for each signer only if it is eligible in the `current` layer

you can't know that without knowing the number of slots eligible in the epoch, and that's why we need to move all of what i mentioned above from the node to the smesher service.

Reply:
i added a commit that basically strips out the struct field in favor of a function scoped map that is passed to `build`. if we want to selectively call build only for eligible signers, we'd still need the restructuring i proposed above.